server/aesdsocket.c

```diff
@@ -0,0 +1,529 @@
+/***********************************************************************
+ * @file      		aesdsocket.c
+ * @version   		0.1
+ * @brief		    Socket server application code
+ *
+ * @author    		Suraj Ajjampur, suraj.ajjampur@Colorado.edu
+ * @date      		Oct 8, 2023
+ *
+ *
+ * @reference
+ *
+ * https://www.thegeekstuff.com/2012/02/c-daemon-process/ 
+ ************************************************************************/
+
+/****************   Includes    ***************/ 
+#include "aesdsocket.h"
+
+/****************   Macros     ***************/ 
+#define DATA_FILE "/var/tmp/aesdsocketdata"
+
+/****************   Global Variables     ***************/ 
+sig_atomic_t fatal_error_in_progress = 0;
+
+// Outout data file
+int dataFileDescriptor;
+
+// Server & Client Socket
+int sock_fd;
+int clientSocketFd;
+
+/****************   Function prototypes     ***************/ 
+void cleanup_on_exit();
+void main_socket_application();
+int run_daemon(void);
+int accept_and_log_client(char *ip_address);
+int receive_and_store_data(void);
+int send_file_content_to_client(void);
+
+
+// Initialize all elements to false
+status_flags s_flags = {false, false, false, false, false, false, false};
+
+/**
+ * @name handle_termination
+ * 
+ * @brief Handler function to cause orderly cleanup or recovery from program error signals and interactive interrupts.
+ * 
+ * The cleanest way for a handler to terminate the process is to raise the same signal that ran the handler
+ * in the first place. This function uses a static variable to keep track of recursive invocation.
+ * 
+ * @param sig Signal number passed to the handler function
+ * 
+ * @source https://www.gnu.org/software/libc/manual/html_node/Termination-in-Handler.html
+ */
+void handle_termination (int sig)
+{
+
+//   Since this handler is established for more than one kind of signal, 
+//      it might still get invoked recursively by delivery of some other kind
+//      of signal.  Use a static variable to keep track of that. 
+  if (fatal_error_in_progress){
+    raise (sig);
+  } 
+  fatal_error_in_progress = 1;
+
+    cleanup_on_exit();
```

Comment:
@Suraj-Ajjampur this strategy needs to ensure cleanup_on_exit is always safe to call from signal handler, which, if you are cleaning up in all scenarios (for instance when you have memory to free) I don't think will be the case.  Think instead about a design which only signals the intent to exit from the signal handler and does cleanup outside the handler itself.

Reply:
Are you referring to scenarios where the struct addrinfo *result, is being allocated memory using the getaddrinfo() function? And when the Termination signals (SIGINT or SIGTERM) is being sent, then it isn't being freed?

One solution I can think of is declaring the *result globally and freeing the memory, in the cleanup_on_exit() function.

Reply:
Created a function 
`void free_and_nullify_result() {
    if (result != NULL) {
        freeaddrinfo(result);
        result = NULL;
    }
}`
And calling in the cleanup function and the in place of freeaddr() in the main_socket_application once result has been used.